vpr/src/noc/read_xml_noc_traffic_flows_file.cpp

```diff
@@ -0,0 +1,230 @@
+    // check that the bandwidth and max latency are positive values
+    if ((traffic_flow_bandwidth < 0) || (max_traffic_flow_latency < 0)) {
+        vpr_throw(VPR_ERROR_OTHER, loc_data.filename_c_str(), loc_data.line(single_flow_tag), "The traffic flow bandwidth and latency constraints need to be positive values.");
+    }
+
+    return;
+}
+
+ClusterBlockId get_router_module_cluster_id(std::string router_module_name, const ClusteringContext& cluster_ctx, pugi::xml_node single_flow_tag, const pugiutil::loc_data& loc_data, const std::vector<ClusterBlockId>& cluster_blocks_compatible_with_noc_router_tiles) {
+    ClusterBlockId router_module_id = ClusterBlockId::INVALID();
+
+    // find the cluster block whos name matches to the provided router module name provided by the user
+    // Then get the corresponding cluster block id if a valid block was found
+    try {
+        router_module_id = cluster_ctx.clb_nlist.find_block_by_name_fragment(router_module_name, cluster_blocks_compatible_with_noc_router_tiles);
+    } catch (const std::regex_error& error) {
+        // if there was an error with matching the regex string,report it to the user here
+        vpr_throw(VPR_ERROR_OTHER, loc_data.filename_c_str(), loc_data.line(single_flow_tag), error.what());
+    }
+
+    // check if a valid block id was found
+    if (router_module_id == ClusterBlockId::INVALID()) {
+        // if here then the module did not exist in the design, so throw an error
+        vpr_throw(VPR_ERROR_OTHER, loc_data.filename_c_str(), loc_data.line(single_flow_tag), "The router module '%s' does not exist in the design.", router_module_name.c_str());
+    }
+
+    return router_module_id;
+}
+
+void check_traffic_flow_router_module_type(std::string router_module_name, ClusterBlockId router_module_id, pugi::xml_node single_flow_tag, const pugiutil::loc_data& loc_data, const ClusteringContext& cluster_ctx, t_physical_tile_type_ptr noc_router_tile_type) {
+    // get the logical type of the provided router module
+    t_logical_block_type_ptr router_module_logical_type = cluster_ctx.clb_nlist.block_type(router_module_id);
+
+    /*
+     * Check whether the current router modules logical type is compatible
+     * with the physical type of a noc router (can the module be placed on a 
+     * noc router tile on the FPGA device). If not then this module is not a router so throw an error.
+     */
+    if (!is_tile_compatible(noc_router_tile_type, router_module_logical_type)) {
+        vpr_throw(VPR_ERROR_OTHER, loc_data.filename_c_str(), loc_data.line(single_flow_tag), "The supplied module name '%s' is not a NoC router.", router_module_name.c_str());
+    }
+
+    return;
+}
+
+t_physical_tile_type_ptr get_physical_type_of_noc_router_tile(const DeviceContext& device_ctx, NocContext& noc_ctx) {
+    // get a reference to a single physical noc router
```

Comment:
Might mention you assume all routers have the same physical type, so you just return the type of the first one. 
If there's any chance that there aren't any routers in the grid then add an assert or error message if physical_noc_router == .end() (should be the test for no router found).

Reply:
Done